ui/src/frontend/index.ts

```diff
@@ -182,26 +166,44 @@ function setupContentSecurityPolicy() {
       'https://*.google.com',
       'https://*.googleusercontent.com',
       'https://www.googletagmanager.com',
-      'https://www.google-analytics.com',
+      'https://*.google-analytics.com',
     ],
     'object-src': ['none'],
-    'connect-src': connectSrc,
+    'connect-src': [
+      `'self'`,
+      'http://127.0.0.1:9001',  // For trace_processor_shell --httpd.
+      'ws://127.0.0.1:9001',    // Ditto, for the websocket RPC.
+      'ws://127.0.0.1:8037',    // For the adb websocket server.
+      'https://*.google-analytics.com',
+      'https://*.googleapis.com',  // For Google Cloud Storage fetches.
+      'blob:',
+      'data:',
+    ],
     'img-src': [
       `'self'`,
       'data:',
       'blob:',
-      'https://www.google-analytics.com',
+      'https://*.google-analytics.com',
       'https://www.googletagmanager.com',
+      'https://*.googleapis.com',
+    ],
+    'style-src': [
+      `'self'`,
+      `'unsafe-inline'`,
     ],
     'navigate-to': ['https://*.perfetto.dev', 'self'],
   };
-  const meta = document.createElement('meta');
-  meta.httpEquiv = 'Content-Security-Policy';
   let policyStr = '';
   for (const [key, list] of Object.entries(policy)) {
     policyStr += `${key} ${list.join(' ')}; `;
   }
-  meta.content = policyStr;
+  return policyStr;
+}
+
+function setupContentSecurityPolicy() {
```

Comment:
nit: it seems a little backwards to me that the Sokatoa app embeds the perfetto timeline into the app's DOM, and that embedded component defines the security policy for the whole app. I guess this solution doesn't look to disrupt that flow but instead allows the app to provide a replacement policy. Ideally, the app would define the policy and set it itself as part of its responsibility to be a secure app, accommodating any needs of any components it embeds (perfetto in this case). E.g., if we were to replace perfetto, we'd end up with no security policy. We'd be defining one, but it wouldn't actually be used! But this is really just a small oddity and not worth worrying about at this point, as I don't think we'll be replacing perfetto any time soon (or at all).

Reply:
You are correct in that I didn't want to disrupt the current behavior but I believe the security policies can only be set on document-level so there is not much we can do. Theia itself does not set any security policies afaik so even if we were to replace Perfetto and somehow overlook the customization code, everything should still work. Besides, we wouldn't even want all the special Google-support in that case ;-) That being said, I think our policy is still to try to modify Perfetto as little as possible to make merging the magical, mystical upstream easier.